src/file/tokio_compat_file.rs

```diff
@@ -513,25 +543,29 @@ impl AsyncWrite for TokioCompatFile {
         let write_limit = self.get_auxiliary().tokio_compat_file_write_limit();
         let mut write_len = self.as_mut().write_len;
 
-        loop {
-            match write_len.checked_add(n as usize) {
-                Some(new_write_len) if new_write_len > write_limit => {}
-                None => {
-                    // case overflow
-                    // This has to be a separate cases since
-                    // write_limit could be set to usize::MAX, in which case
-                    // saturating_add would never return anything larger than it.
-                }
-                Some(new_write_len) => {
-                    *self.as_mut().project().write_len = new_write_len;
-                    break;
-                }
-            }
-
+        if write_len == write_limit {
             ready!(self.as_mut().flush_one(cx))?;
             write_len = self.as_mut().write_len;
         }
 
+        match write_len.checked_add(n as usize) {
+            Some(new_write_len) if new_write_len > write_limit => {
+                n = (write_limit - write_len) as u32;
+                *self.as_mut().project().write_len = write_limit;
```

Comment:
We should increment write_len after the future is pushed, since the sending step could also fail (due to serialisation error for now, but maybe more in the future).